### PR TITLE
Extract collection-expression creation code into its own helper type.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This code will be used by other rewriting features being worked on as part of https://github.com/dotnet/roslyn/issues/69132.

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+    /// </summary>
+    public static async Task<CollectionExpressionSyntax> CreateCollectionExpressionAsync<TParentExpression>(
+        Document workspaceDocument,
+        CodeActionOptionsProvider fallbackOptions,
+        TParentExpression expressionToReplace,
+        ImmutableArray<CollectionExpressionMatch> matches,
+        Func<TParentExpression, InitializerExpressionSyntax?> getInitializer,
+        Func<TParentExpression, InitializerExpressionSyntax, TParentExpression> withInitializer,
+        CancellationToken cancellationToken)
+        where TParentExpression : ExpressionSyntax
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs:42]: this is the same as the original code, just updated to not have to take an ObjectCreationExpression as the parent.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpCollectionInitializerRewriter.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer;
+
+using static SyntaxFactory;
+
+internal static class CSharpCollectionInitializerRewriter
+{
+    public static ExpressionSyntax ConvertExpression(
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpCollectionInitializerRewriter.cs:16]: these are all moves, unchanged.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider_CollectionExpression.cs [hunk 3]
```diff
@@ -2,588 +2,38 @@
...
-            static StatementSyntax UnwrapEmbeddedStatement(StatementSyntax statement)
-                => statement is BlockSyntax { Statements: [var innerStatement] } ? innerStatement : statement;
+            return CSharpCollectionExpressionRewriter.CreateCollectionExpressionAsync(
+                document,
+                fallbackOptions,
+                objectCreation,
+                matches.SelectAsArray(m => new CollectionExpressionMatch(m.Statement, m.UseSpread)),
+                o => o.Initializer,
+                (o, i) => o.WithInitializer(i),
+                cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider_CollectionExpression.cs:36]: exactly the same as before, except:

1. Converting this feature's 'Match' type to 'CollectionExpressionMatch'
2. Passing the lambdas to make this code work with ObjectCreationExpressions

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal.  Very simple refactoring here that helps the next PR.  Thanks!

File: src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider_CollectionExpression.cs [hunk 4]
```diff
@@ -2,588 +2,38 @@
...
-
-            static StatementSyntax UnwrapEmbeddedStatement(StatementSyntax statement)
-                => statement is BlockSyntax { Statements: [var innerStatement] } ? innerStatement : statement;
+            return CSharpCollectionExpressionRewriter.CreateCollectionExpressionAsync(
+                document,
+                fallbackOptions,
+                objectCreation,
+                matches.SelectAsArray(m => new CollectionExpressionMatch(m.Statement, m.UseSpread)),
+                o => o.Initializer,
+                (o, i) => o.WithInitializer(i),
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider_CollectionExpression.cs:35]:
```suggestion
                static objectCreation => objectCreation.Initializer,
                static (objectCreation, initializer) => objectCreation.WithInitializer(initializer),
```

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne @akhera99 ptal :)

**Review by akhera99 — APPROVED** [review]

**Pull request merged**